pkg/lib/operatorstatus/monitor.go

```diff
@@ -0,0 +1,221 @@
+package operatorstatus
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	configv1client "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/operatorclient"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// Wait time before we probe next while checking whether cluster
+	// operator API is available.
+	defaultProbeInterval = 1 * time.Minute
+
+	// Default size of the notification channel.
+	defaultNotificationChannelSize = 64
+)
+
+// NewMonitor returns a new instance of Monitor that can be used to continuously
+// update a clusteroperator resource and an instance of Sender that can be used
+// to send update notifications to it.
+//
+// The name of the clusteroperator resource to monitor is specified in name.
+func NewMonitor(name string, log *logrus.Logger, opClient operatorclient.ClientInterface, configClient configv1client.ConfigV1Interface) (Monitor, Sender) {
+	logger := log.WithField("monitor", "clusteroperator")
+	discovery := opClient.KubernetesInterface().Discovery()
+	releaseVersion := os.Getenv("RELEASE_VERSION")
+	names := split(name)
+
+	logger.Infof("monitoring the following components %s", names)
+
+	monitor := &monitor{
+		logger:         logger,
+		reporter:       newReporter(releaseVersion),
+		writer:         NewWriter(discovery, configClient),
+		notificationCh: make(chan *NotificationContext, defaultNotificationChannelSize),
+		names:          names,
+	}
+
+	return monitor, monitor
+}
+
+// NotificationContext contains all necessary information related to a notification.
+type NotificationContext struct {
+	// Name of the clusteroperator resource associated with this CSV.
+	Name string
+
+	// Current is the CSV for which we have received notification.
+	Current *v1alpha1.ClusterServiceVersion
+
+	// If there is an upgrade going on, Final is set to the latest version of
+	// the CSV that is replacing current.
+	// For a chain like this, Current (v1) -> v2 -> v3 -> Final (v4)
+	// Final will be set to the CSV linked to v4.
+	// Otherwise this value is set to nil.
+	Final *v1alpha1.ClusterServiceVersion
+
+	// Deleted indicates whether Current has been deleted.
+	Deleted bool
+}
+
+// Monitor is an interface that wraps the Run method.
+//
+// Run runs forever, it reads from an underlying notification channel and
+// updates an clusteroperator resource accordingly.
+// If the specified stop channel is closed the loop must terminate gracefully.
+type Monitor interface {
+	Run(stopCh <-chan struct{})
```

Comment:
I wonder if we want to take a [context.Context](https://golang.org/pkg/context/) here instead.

Reply:
`main.go` can introduce a parent context and that parent context can be passed into this `Run` method. I think one of your PR(s) introduces a parent context in `main.go`. Once that goes in, I can then use it.